src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs

```diff
@@ -26,7 +26,7 @@ namespace Microsoft.CodeAnalysis.Diagnostics
     internal partial class AnalyzerManager
     {
         // This cache stores the analyzer execution context per-analyzer (i.e. registered actions, supported descriptors, etc.).
-        private readonly ImmutableDictionary<DiagnosticAnalyzer, AnalyzerExecutionContext> _analyzerExecutionContextMap;
+        private readonly Dictionary<DiagnosticAnalyzer, AnalyzerExecutionContext> _analyzerExecutionContextMap;
```

Comment:
doc that it is effectively immutable, but is kept as a Dictionary for perf.

Reply:
Given we don't mutate after creation could we store here is `IReadOnlyDictionary`? That would help enforce the "don't mutate" part without changing lookup characteristics.

Reply:
I considered that, but wouldn't that make each access a bit slower because of the virtual call? The primary intent here was to speed up those calls.

Reply:
Did you measure that? The underlying issue this PR is solving is an algorithmic one: hash map vs. AVL tree lookup. Doubting that the extra indirect of a virtual call is going to impact this too much. Particularly cause we use interfaces everywhere in Roslyn.

Reply:
I have not measured the performance implications of declaring the type as IReadOnlyDictionary vs Dictionary. I was just trying to make the simplest changes to make as big of a performance impact as possible in this codepath. 

If you feel fairly strongly, I have no issue with making it IReadOnlyDictionary and if prism flags it in the future we can reconsider changing the type declaration to Dictionary. (I doubt it would ever get flagged for just the vcall overhead though)

Reply:
my preference is Dictionary.  I think it's fine whern encapsulated.

Reply:
@jaredpar -- does this sound ok for you? Anything else of concern?